### ISSUE TITLE
What can cause a reserved job to not be ready?

### ISSUE THREAD

**Issue by georaldc** [beanstalkd/beanstalkd#659]
Ran into a strange issue that took down our application. Basically, our beanstalkd server (running the latest available version, 1.13) suddenly became unusable and we started seeing thousands of error logs that were pointing to this specific line

https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/prot.c#L466

I believe this is the first time we've encountered this issue. Other issues we've had is we would get random segmentation faults and that seems to happen once every few weeks. I'm not sure if they are related to this new issue (I'm still hoping to capture a coredump for the segfault problem) but this is the first time we've seen this particular problem occur after years of use and it's a bit worse because it prevents usage of the queue server altogether (at least with the segfaults, our beanstalkd server process gets restarted immediately and resumes work from binlogs). Restarting the process by hand fixed the issue immediately.

**Comment by juhasuni** [comment]
@georaldc did you find out why this happens? We encountered the same issue but now it doesn't seem to occur any more as we configured Beanstalkd to write the bin file to local disk on our AWS Fargate instance instead of using mounted EFS storage.

**Comment by juhasuni** [comment]
So I'm saying it might have something to do with the disk storage you're using. We still need to monitor this more to see if the problem is actually resolved. For us this is not ideal solution as we still need to permanently store the binlog files to EFS or else they will disappear when the Fargate instance is restarted.

**Comment by georaldc** [comment]
@juhasuni Not yet. We also write to local disk for binlogs so I don't think that's related. We do run beanstalkd on a server that also doubles as a web server with random spikes to load, so that might exacerbate things. We'll likely plan on moving beanstalkd into its own server in the near future to see if that will help.

**Comment by georaldc** [comment]
This issue happened again to us just now, almost a month after it previously occurred (so beanstalkd has been running for 30+ days without issue until it started complaining again about reserved jobs not being in a ready state yesterday). It's almost like it something slowly builds up and leads to this issue. Anyone have any ideas on what might cause this?

**Comment by georaldc** [comment]
I'm not familiar with C code, so I apologize if what I'm saying doesn't make sense, but could this be some sort of race condition when it comes to delayed jobs being promoted to a Ready state? The error seems to suggest that a job retrieved from a tube's ready heap doesn't actually have a Ready r.state value and existing Delayed jobs get inserted into this data structure via enqueue_job

https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/prot.c#L2161
https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/prot.c#L519

But the job's state gets modified only after heapinsert is called

https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/prot.c#L522

**Comment by georaldc** [comment]
Maybe related to this, but my application somehow was able to insert a single job (out of thousands of jobs) into a tube with truncated data. This caused all the clients I rely on (PHP, so pheanstalk for reading/writing jobs and beanstalk_console for a quick web-based management interface) to complain about a reserved job missing CRLF bytes, causing the job to be returned into its tube in a ready state again, effectively locking up workers watching that tube permanently. The only fix was for me to connect directly into beanstalkd and delete the offending job.

Interestingly, a few minutes after deleting the job, my beanstalkd process crashed with a "segfault, error 4" error, which is something I have been noticing happening almost once every month now. Not much info, but it again feels my beanstalkd process slowly becomes unstable after several days of running until it starts manifesting weird bugs like this.

**Comment by silviucpp** [comment]
I'm using beanstalkd for 10 years and queue/consume in it over 1M jobs daily (persistence on disk is disabled) and I never experienced such an issue. I use a slightly modified version of the server but is based on the last version of beanstalkd from master. maybe you should check somehow the clients you have or try to isolate better the payload you sent and to replicate the problem. beanstalkd as a server is very stable and without detailed information is hard to get help.

**Comment by ysmolski** [comment]
@georaldc This might be an issue with the client itself. If your client is unstable then it will be sending malformed data to beanstalkd and that could lead to rare bug... The only way to fix such bug, is to reproduce the case, I am afraid.
